File: projects/python-labs/lab-08-control-flow/exercise-02-pattern-matching.py
from dataclasses import dataclass

@dataclass
class SimpleTrade:
    side: str
    quantity: int


def classify_trade(trade: SimpleTrade) -> str:
    match trade:
        case SimpleTrade(side="BUY", quantity=q) if q > 500:
            return "large-buy"
        case SimpleTrade(side="SELL", quantity=q) if q > 500:
            return "large-sell"
        case SimpleTrade(side="BUY", quantity=q) if q <= 500:
            return "small-buy"
        case SimpleTrade(side="SELL", quantity=q) if q <= 500:
            return "small-sell"
        case _:
            return "invalid"


def test_classify_trade() -> None:
    trades = [
        SimpleTrade("BUY", 100),    # small-buy
        SimpleTrade("BUY", 900),    # large-buy
        SimpleTrade("SELL", 50),    # small-sell
        SimpleTrade("SELL", 700),   # large-sell
        SimpleTrade("HOLD", 100),   # invalid
        SimpleTrade("BUY", -1),     # small-buy by raw rule; still matches <= 500
    ]

    for t in trades:
        print(t, "->", classify_trade(t))


test_classify_trade()

File: projects/csharp-labs/lab-08-pattern-matching/exercise-02-pattern-matching.cs
using System;

public record Trade(string Side, int Quantity);

public static class TradeClassifier
{
    public static string ClassifyTrade(Trade trade) =>
        trade switch
        {
            { Side: "BUY", Quantity: > 500 }  => "large-buy",
            { Side: "SELL", Quantity: > 500 } => "large-sell",
            { Side: "BUY", Quantity: <= 500 } => "small-buy",
            { Side: "SELL", Quantity: <= 500 } => "small-sell",
            _ => "invalid"
        };

    public static void Main()
    {
        var trades = new[]
        {
            new Trade("BUY", 100),
            new Trade("BUY", 900),
            new Trade("SELL", 50),
            new Trade("SELL", 700),
            new Trade("HOLD", 100),
            new Trade("BUY", -1)
        };

        foreach (var trade in trades)
        {
            Console.WriteLine($"{trade} -> {ClassifyTrade(trade)}");
        }
    }
}